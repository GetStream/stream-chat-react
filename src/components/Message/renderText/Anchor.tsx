import clsx from 'clsx';
import React, { ComponentProps } from 'react';
import { ReactMarkdownProps } from 'react-markdown/lib/complex-types';

export const Anchor = ({ children, href }: ComponentProps<'a'> & ReactMarkdownProps) => {
  const isEmail = href?.startsWith('mailto:');
  const isUrl = href?.startsWith('http');

  if (!href || (!isEmail && !isUrl)) return <>{children}</>;

  return (
    <a
      className={clsx({ 'str-chat__message-url-link': isUrl })}
      href={href}
      rel='nofollow noreferrer noopener'
      target='_blank'
    >
      {children}
    </a>
  );
};
