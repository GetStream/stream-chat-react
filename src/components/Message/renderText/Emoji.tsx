import React from 'react';
import { ReactMarkdownProps } from 'react-markdown/lib/complex-types';

export const Emoji = ({ children }: ReactMarkdownProps) => (
  <span className='inline-text-emoji' data-testid='inline-text-emoji'>
    {children}
  </span>
);
