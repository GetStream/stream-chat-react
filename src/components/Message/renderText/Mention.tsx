import React from 'react';

import type { ReactMarkdownProps } from 'react-markdown/lib/complex-types';
import type { UserResponse } from 'stream-chat';
import type { DefaultStreamChatGenerics } from '../../../types/types';

export type MentionProps<
  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics
> = ReactMarkdownProps & {
  /**
   * @deprecated will be removed in the next major release, transition to using `node.mentionedUser` instead
   */
  mentioned_user: UserResponse<StreamChatGenerics>;
  node: {
    /**
     * @deprecated will be removed in the next major release, transition to using `node.mentionedUser` instead
     */
    mentioned_user: UserResponse<StreamChatGenerics>;
    mentionedUser: UserResponse<StreamChatGenerics>;
  };
};
export const Mention = <
  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics
>({
  children,
  node: { mentionedUser },
}: MentionProps<StreamChatGenerics>) => (
  <span className='str-chat__message-mention' data-user-id={mentionedUser.id}>
    {children}
  </span>
);
