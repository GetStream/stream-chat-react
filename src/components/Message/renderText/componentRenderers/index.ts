export * from './Anchor';
export * from './Emoji';
export * from './Mention';
