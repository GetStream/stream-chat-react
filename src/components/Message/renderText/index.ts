export type { MentionProps } from './componentRenderers';
export { escapeRegExp, matchMarkdownLinks, messageCodeBlocks } from './regex';
export * from './rehypePlugins';
export * from './remarkPlugins';
export * from './renderText';
