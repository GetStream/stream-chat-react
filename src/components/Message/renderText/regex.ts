export function escapeRegExp(text: string) {
  return text.replace(/[-[\]{}()*+?.,/\\^$|#]/g, '\\$&');
}

export const detectHttp = /(http(s?):\/\/)?(www\.)?/;

export const messageCodeBlocks = (message: string) => {
  const codeRegex = /```[a-z]*\n[\s\S]*?\n```|`[a-z]*[\s\S]*?`/gm;
  const matches = message.match(codeRegex);
  return matches || [];
};

export const matchMarkdownLinks = (message: string) => {
  const regexMdLinks = /\[([^[]+)\](\(.*\))/gm;
  const matches = message.match(regexMdLinks);
  const singleMatch = /\[([^[]+)\]\((.*)\)/;

  const links = matches
    ? matches.map((match) => {
        const i = singleMatch.exec(match);
        return i && [i[1], i[2]];
      })
    : [];

  return links.flat();
};
