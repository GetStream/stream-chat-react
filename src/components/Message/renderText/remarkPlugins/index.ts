export * from './htmlToTextPlugin';
export * from './keepLineBreaksPlugin';
