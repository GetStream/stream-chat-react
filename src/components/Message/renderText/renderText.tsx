import React, { ComponentType } from 'react';
import ReactMarkdown, { defaultUrlTransform, Options } from 'react-markdown';
import { find } from 'linkifyjs';
import uniqBy from 'lodash.uniqby';
import remarkGfm from 'remark-gfm';

import { Anchor, Emoji, Mention, MentionProps } from './componentRenderers';
import { detectHttp, escapeRegExp, matchMarkdownLinks, messageCodeBlocks } from './regex';
import { emojiMarkdownPlugin, mentionsMarkdownPlugin } from './rehypePlugins';
import { htmlToTextPlugin, keepLineBreaksPlugin } from './remarkPlugins';

import type { PluggableList } from 'react-markdown/lib';
import type { UserResponse } from 'stream-chat';
import type { DefaultStreamChatGenerics } from '../../../types/types';

export type RenderTextPluginConfigurator = (defaultPlugins: PluggableList) => PluggableList;

export const defaultAllowedTagNames: Array<keyof JSX.IntrinsicElements | 'emoji' | 'mention'> = [
  'html',
  'text',
  'br',
  'p',
  'em',
  'strong',
  'a',
  'ol',
  'ul',
  'li',
  'code',
  'pre',
  'blockquote',
  'del',
  // custom types (tagNames)
  'emoji',
  'mention',
];

function formatUrlForDisplay(url: string) {
  try {
    return decodeURIComponent(url).replace(detectHttp, '');
  } catch (e) {
    return url;
  }
}

function encodeDecode(url: string) {
  try {
    return encodeURI(decodeURIComponent(url));
  } catch (error) {
    return url;
  }
}

const urlTransform = (uri: string) => (uri.startsWith('app://') ? uri : defaultUrlTransform(uri));

const getPluginsForward: RenderTextPluginConfigurator = (plugins: PluggableList) => plugins;

export const markDownRenderers: RenderTextOptions['customMarkDownRenderers'] = {
  a: Anchor,
  emoji: Emoji,
  mention: Mention,
};

export type RenderTextOptions<
  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics
> = {
  // eslint-disable-next-line @typescript-eslint/ban-types
  allowedTagNames?: Array<keyof JSX.IntrinsicElements | 'emoji' | 'mention' | (string & {})>;
  customMarkDownRenderers?: Options['components'] &
    Partial<{
      emoji: ComponentType;
      mention: ComponentType<MentionProps<StreamChatGenerics>>;
    }>;
  getRehypePlugins?: RenderTextPluginConfigurator;
  getRemarkPlugins?: RenderTextPluginConfigurator;
};

export const renderText = <
  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics
>(
  text?: string,
  mentionedUsers?: UserResponse<StreamChatGenerics>[],
  {
    allowedTagNames = defaultAllowedTagNames,
    customMarkDownRenderers,
    getRehypePlugins = getPluginsForward,
    getRemarkPlugins = getPluginsForward,
  }: RenderTextOptions = {},
) => {
  // take the @ mentions and turn them into markdown?
  // translate links
  if (!text) return null;
  if (text.trim().length === 1) return <>{text}</>;

  let newText = text;
  const markdownLinks = matchMarkdownLinks(newText);
  const codeBlocks = messageCodeBlocks(newText);

  // extract all valid links/emails within text and replace it with proper markup
  uniqBy([...find(newText, 'email'), ...find(newText, 'url')], 'value').forEach(
    ({ href, type, value }) => {
      const linkIsInBlock = codeBlocks.some((block) => block?.includes(value));

      // check if message is already  markdown
      const noParsingNeeded =
        markdownLinks &&
        markdownLinks.filter((text) => {
          const strippedHref = href?.replace(detectHttp, '');
          const strippedText = text?.replace(detectHttp, '');

          if (!strippedHref || !strippedText) return false;

          return strippedHref.includes(strippedText) || strippedText.includes(strippedHref);
        });

      if (noParsingNeeded.length > 0 || linkIsInBlock) return;

      try {
        // special case for mentions:
        // it could happen that a user's name matches with an e-mail format pattern.
        // in that case, we check whether the found e-mail is actually a mention
        // by naively checking for an existence of @ sign in front of it.
        if (type === 'email' && mentionedUsers) {
          const emailMatchesWithName = mentionedUsers.some((u) => u.name === value);
          if (emailMatchesWithName) {
            newText = newText.replace(new RegExp(escapeRegExp(value), 'g'), (match, position) => {
              const isMention = newText.charAt(position - 1) === '@';
              // in case of mention, we leave the match in its original form,
              // and we let `mentionsMarkdownPlugin` to do its job
              return isMention ? match : `[${match}](${encodeDecode(href)})`;
            });

            return;
          }
        }

        const displayLink = type === 'email' ? value : formatUrlForDisplay(href);

        newText = newText.replace(
          new RegExp(escapeRegExp(value), 'g'),
          `[${displayLink}](${encodeDecode(href)})`,
        );
      } catch (e) {
        void e;
      }
    },
  );

  const remarkPlugins: PluggableList = [
    htmlToTextPlugin,
    keepLineBreaksPlugin,
    [remarkGfm, { singleTilde: false }],
  ];
  const rehypePlugins: PluggableList = [emojiMarkdownPlugin];

  if (mentionedUsers?.length) {
    rehypePlugins.push(mentionsMarkdownPlugin(mentionedUsers));
  }

  return (
    <ReactMarkdown
      allowedElements={allowedTagNames}
      components={{
        ...markDownRenderers,
        ...customMarkDownRenderers,
      }}
      rehypePlugins={getRehypePlugins(rehypePlugins)}
      remarkPlugins={getRemarkPlugins(remarkPlugins)}
      skipHtml
      unwrapDisallowed
      urlTransform={urlTransform}
    >
      {newText}
    </ReactMarkdown>
  );
};
