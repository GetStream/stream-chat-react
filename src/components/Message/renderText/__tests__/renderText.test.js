import React from 'react';
import renderer from 'react-test-renderer';
import { findAndReplace } from 'hast-util-find-and-replace';
import { u } from 'unist-builder';
import { htmlToTextPlugin, keepLineBreaksPlugin } from '../remarkPlugins';
import { defaultAllowedTagNames, renderText } from '../renderText';

const strikeThroughText = '~~xxx~~';

describe(`renderText`, () => {
  it('handles the special case where user name matches to an e-mail pattern - 1', () => {
    const Markdown = renderText(
      'Hello @username@email.com, is username@email.com your @primary e-mail?',
      [{ id: 'id-username@email.com', name: 'username@email.com' }],
    );
    const tree = renderer.create(Markdown).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('handles the special case where user name matches to an e-mail pattern - 2', () => {
    const Markdown = renderText(
      'username@email.com @username@email.com is this the right address?',
      [{ id: 'id-username@email.com', name: 'username@email.com' }],
    );
    const tree = renderer.create(Markdown).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('handles the special case where user name matches to an e-mail pattern - 3', () => {
    const Markdown = renderText(
      '@username@email.com @username@email.com @username@email.com @username@email.com',
      [{ id: 'id-username@email.com', name: 'username@email.com' }],
    );
    const tree = renderer.create(Markdown).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('handles the special case where user name matches to an e-mail pattern - 4', () => {
    const Markdown = renderText(
      '@username@email.com @username@email.com username@email.com @username@email.com',
      [{ id: 'id-username@email.com', name: 'username@email.com' }],
    );
    const tree = renderer.create(Markdown).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders custom mention', () => {
    const CustomMention = (props) => (
      <span className='my-mention' data-node-mentionedUser-id={props.node.mentionedUser.id}>
        {props.children}
      </span>
    );
    const Markdown = renderText(
      '@username@email.com @username@email.com username@email.com @username@email.com',
      [{ id: 'id-username@email.com', name: 'username@email.com' }],
      {
        customMarkDownRenderers: {
          mention: CustomMention,
        },
      },
    );
    const tree = renderer.create(Markdown).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders standard markdown text', () => {
    const Markdown = renderText('Hi, shall we meet on **Tuesday**?', []);
    const tree = renderer.create(Markdown).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders remark-gfm list and strikethrough correctly', () => {
    const Markdown = renderText(
      'Pick a time to meet:\n- Wednesday\n- Thursday\n- ~~Sunday~~\n- ~Monday~\n',
      [],
    );
    const tree = renderer.create(Markdown).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("handles the special case where there's at least one mention and @ symbol at the end", () => {
    const Markdown = renderText('@username@email.com @', [
      { id: 'id-username@email.com', name: 'username@email.com' },
    ]);
    const tree = renderer.create(Markdown).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('handles the special case where there are pronouns in the name', () => {
    const Markdown = renderText('hey, @John (they/them), how are you?', [
      { id: 'john', name: 'John (they/them)' },
    ]);
    const tree = renderer.create(Markdown).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('handles the special case where there is a forward slash in the name', () => {
    const Markdown = renderText('hey, @John/Cena, how are you?', [
      { id: 'john', name: 'John/Cena' },
    ]);
    const tree = renderer.create(Markdown).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('handles the special case where there is a backslash in the name', () => {
    const Markdown = renderText('hey, @John\\Cena, how are you?', [
      { id: 'john', name: 'John\\Cena' },
    ]);
    const tree = renderer.create(Markdown).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('parses user mention to default format', () => {
    const Markdown = renderText('@username@email.com', [
      { id: 'id-username@email.com', name: 'username@email.com' },
    ]);
    const tree = renderer.create(Markdown).toJSON();
    expect(tree).toMatchInlineSnapshot(`
      <p>
        <span
          className="str-chat__message-mention"
          data-user-id="id-username@email.com"
        >
          @username@email.com
        </span>
      </p>
    `);
  });

  it('allows to override rehype plugins', () => {
    const customPlugin = () => (tree) => tree;
    const getRehypePlugins = () => [customPlugin];
    const Markdown = renderText(
      '@username@email.com',
      [{ id: 'id-username@email.com', name: 'username@email.com' }],
      { getRehypePlugins },
    );
    const tree = renderer.create(Markdown).toJSON();
    expect(tree).toMatchInlineSnapshot(`
      <p>
        @
        <a
          className=""
          href="mailto:username@email.com"
          rel="nofollow noreferrer noopener"
          target="_blank"
        >
          username@email.com
        </a>
      </p>
    `);
  });

  it('allows to merge custom rehype plugins followed by default rehype plugins', () => {
    const customPlugin = () => (tree) => findAndReplace(tree, [/.*@.*/, () => u('text', '#')]);
    const getRehypePlugins = (defaultPlugins) => [customPlugin, ...defaultPlugins];
    const Markdown = renderText(
      '@username@email.com',
      [{ id: 'id-username@email.com', name: 'username@email.com' }],
      { getRehypePlugins },
    );
    const tree = renderer.create(Markdown).toJSON();
    expect(tree).toMatchInlineSnapshot(`
      <p>
        #
        <a
          className=""
          href="mailto:username@email.com"
          rel="nofollow noreferrer noopener"
          target="_blank"
        >
          #
        </a>
      </p>
    `);
  });

  it('allows to merge default rehype plugins followed by custom rehype plugins', () => {
    const customPlugin = () => (tree) => findAndReplace(tree, [/.*@.*/, () => u('text', '#')]);
    const getRehypePlugins = (defaultPlugins) => [...defaultPlugins, customPlugin];
    const Markdown = renderText(
      '@username@email.com',
      [{ id: 'id-username@email.com', name: 'username@email.com' }],
      { getRehypePlugins },
    );
    const tree = renderer.create(Markdown).toJSON();
    expect(tree).toMatchInlineSnapshot(`
      <p>
        <span
          className="str-chat__message-mention"
          data-user-id="id-username@email.com"
        >
          #
        </span>
      </p>
    `);
  });

  it('renders strikethrough', () => {
    const Markdown = renderText(strikeThroughText);
    const tree = renderer.create(Markdown).toJSON();
    expect(tree).toMatchInlineSnapshot(`
      <p>
        <del>
          xxx
        </del>
      </p>
    `);
  });

  it('allows to override remark plugins', () => {
    const customPlugin = () => (tree) => tree;
    const getRemarkPlugins = () => [customPlugin];
    const Markdown = renderText(strikeThroughText, [], { getRemarkPlugins });
    const tree = renderer.create(Markdown).toJSON();
    expect(tree).toMatchInlineSnapshot(`
      <p>
        ~~xxx~~
      </p>
    `);
  });

  it('executes remark-gfm before the custom remark plugins are executed', () => {
    const replace = () => u('text', '#');
    const customPlugin = () => (tree) =>
      findAndReplace(tree, [new RegExp(strikeThroughText), replace]);

    const getRemarkPluginsFirstCustom = (defaultPlugins) => [customPlugin, ...defaultPlugins];
    const getRemarkPluginsFirstDefault = (defaultPlugins) => [...defaultPlugins, customPlugin];
    [getRemarkPluginsFirstCustom, getRemarkPluginsFirstDefault].forEach((getRemarkPlugins) => {
      const Markdown = renderText(strikeThroughText, [], {
        getRemarkPlugins,
      });
      const tree = renderer.create(Markdown).toJSON();
      expect(tree).toMatchInlineSnapshot(`
              <p>
                <del>
                  xxx
                </del>
              </p>
          `);
    });
  });

  it('allows to render custom elements', () => {
    const customTagName = 'xxx';
    const text = 'a b c';
    const replace = (match) =>
      u('element', { properties: {}, tagName: customTagName }, [u('text', match)]);
    const customPlugin = () => (tree) => findAndReplace(tree, [/b/, replace]);
    const getRehypePlugins = (defaultPlugins) => [customPlugin, ...defaultPlugins];
    const Markdown = renderText(text, [], {
      allowedTagNames: [...defaultAllowedTagNames, customTagName],
      getRehypePlugins,
    });
    const tree = renderer.create(Markdown).toJSON();
    expect(tree).toMatchInlineSnapshot(`
      <p>
        a 
        <xxx>
          b
        </xxx>
         c
      </p>
    `);
  });
});

describe('keepLineBreaksPlugin', () => {
  const lineBreaks = '\n\n\n';
  const paragraphText = `a${lineBreaks}b${lineBreaks}c`;
  const unorderedListText = `* item 1${lineBreaks}* item 2${lineBreaks}* item 3`;
  const orderedListText = `1. item 1${lineBreaks}2. item 2${lineBreaks}3. item 3`;
  const headingText = `## Heading${lineBreaks}a`;
  const codeBlockText = 'a\n\n\n```b```\n\n\nc';
  const horizontalRuleText = `a${lineBreaks}---${lineBreaks}b`;
  const blockquoteText = `a${lineBreaks}>b${lineBreaks}c`;
  const withStrikeThroughText = `a${lineBreaks}${strikeThroughText}${lineBreaks}b`;
  const tableText = `a${lineBreaks}| a | b  |  c |  d  |\n| - | :- | -: | :-: |\n| a | b  |  c |  d  |${lineBreaks}c`;

  const doRenderText = (text, present) => {
    const Markdown = renderText(
      text,
      {},
      { getRemarkPlugins: () => (present ? [keepLineBreaksPlugin] : []) },
    );
    return renderer.create(Markdown).toJSON();
  };

  describe('absent', () => {
    const present = false;
    it(`does not keep line breaks between paragraphs`, () => {
      const tree = doRenderText(paragraphText, present);
      expect(tree).toMatchSnapshot();
    });
    it(`does not keep line breaks between the items in an unordered list`, () => {
      const tree = doRenderText(unorderedListText, present);
      expect(tree).toMatchSnapshot();
    });
    it(`does not keep line breaks between the items in an ordered list`, () => {
      const tree = doRenderText(orderedListText, present);
      expect(tree).toMatchSnapshot();
    });
    it(`does not keep line breaks under a heading`, () => {
      const tree = doRenderText(headingText, present);
      expect(tree).toMatchSnapshot();
    });
    it(`does not keep line breaks around a horizontal rule`, () => {
      const tree = doRenderText(horizontalRuleText, present);
      expect(tree).toMatchSnapshot();
    });
    it(`does not keep line breaks around a code block`, () => {
      const tree = doRenderText(codeBlockText, present);
      expect(tree).toMatchSnapshot();
    });
    it(`does not keep line breaks around a blockquote`, () => {
      const tree = doRenderText(blockquoteText, present);
      expect(tree).toMatchSnapshot();
    });
    it(`does not keep line breaks around a strikethrough`, () => {
      const tree = doRenderText(withStrikeThroughText, present);
      expect(tree).toMatchSnapshot();
    });
    it(`does not keep line breaks around a table`, () => {
      const tree = doRenderText(tableText, present);
      expect(tree).toMatchSnapshot();
    });
  });
  describe('present', () => {
    const present = true;
    it(`keeps line breaks between paragraphs`, () => {
      const tree = doRenderText(paragraphText, present);
      expect(tree).toMatchSnapshot();
    });
    it(`keeps line breaks between the items in an unordered list`, () => {
      const tree = doRenderText(unorderedListText, present);
      expect(tree).toMatchSnapshot();
    });
    it(`keeps line breaks between the items in an ordered list`, () => {
      const tree = doRenderText(orderedListText, present);
      expect(tree).toMatchSnapshot();
    });
    it(`keeps line breaks under a heading`, () => {
      const tree = doRenderText(headingText, present);
      expect(tree).toMatchSnapshot();
    });
    it(`keeps line breaks around a horizontal rule`, () => {
      const tree = doRenderText(horizontalRuleText, present);
      expect(tree).toMatchSnapshot();
    });
    it(`keeps line breaks around a code block`, () => {
      const tree = doRenderText(codeBlockText, present);
      expect(tree).toMatchSnapshot();
    });
    it(`keeps line breaks around a blockquote`, () => {
      const tree = doRenderText(blockquoteText, present);
      expect(tree).toMatchSnapshot();
    });

    it(`keeps line breaks around a strikethrough`, () => {
      const tree = doRenderText(withStrikeThroughText, present);
      expect(tree).toMatchSnapshot();
    });
    it(`keeps line breaks around a table`, () => {
      const tree = doRenderText(tableText, present);
      expect(tree).toMatchSnapshot();
    });
  });
});

describe('htmlToTextPlugin', () => {
  const renderTextWithHtml = (withPlugin) => {
    const textWithHtml = `
<div>
  <script>console.error('This error should not be logged from renderText.test.js!')</script>
</div>
`;
    const Markdown = renderText(
      textWithHtml,
      {},
      { getRemarkPlugins: () => (withPlugin ? [htmlToTextPlugin] : []) },
    );
    return renderer.create(Markdown).toJSON();
  };

  it(`absent does not keep HTML in text`, () => {
    const tree = renderTextWithHtml(false);
    expect(tree).toMatchSnapshot();
  });

  it(`present keeps HTML in text`, () => {
    const tree = renderTextWithHtml(true);
    expect(tree).toMatchSnapshot();
  });
});
