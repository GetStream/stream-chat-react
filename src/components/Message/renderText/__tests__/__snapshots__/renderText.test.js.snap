// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`htmlToTextPlugin absent does not keep HTML in text 1`] = `null`;

exports[`htmlToTextPlugin present keeps HTML in text 1`] = `
"<div>
<script>console.error('This error should not be logged from renderText.test.js!')</script>
</div>"
`;

exports[`keepLineBreaksPlugin absent does not keep line breaks around a blockquote 1`] = `
Array [
  <p>
    a
  </p>,
  "
",
  <blockquote>
    

    <p>
      b
    </p>
    

  </blockquote>,
  "
",
  <p>
    c
  </p>,
]
`;

exports[`keepLineBreaksPlugin absent does not keep line breaks around a code block 1`] = `
Array [
  <p>
    a
  </p>,
  "
",
  <p>
    <code>
      b
    </code>
  </p>,
  "
",
  <p>
    c
  </p>,
]
`;

exports[`keepLineBreaksPlugin absent does not keep line breaks around a horizontal rule 1`] = `
Array [
  <p>
    a
  </p>,
  "
",
  "
",
  <p>
    b
  </p>,
]
`;

exports[`keepLineBreaksPlugin absent does not keep line breaks around a strikethrough 1`] = `
Array [
  <p>
    a
  </p>,
  "
",
  <p>
    ~~xxx~~
  </p>,
  "
",
  <p>
    b
  </p>,
]
`;

exports[`keepLineBreaksPlugin absent does not keep line breaks around a table 1`] = `
Array [
  <p>
    a
  </p>,
  "
",
  <p>
    | a | b  |  c |  d  |
| - | :- | -: | :-: |
| a | b  |  c |  d  |
  </p>,
  "
",
  <p>
    c
  </p>,
]
`;

exports[`keepLineBreaksPlugin absent does not keep line breaks between paragraphs 1`] = `
Array [
  <p>
    a
  </p>,
  "
",
  <p>
    b
  </p>,
  "
",
  <p>
    c
  </p>,
]
`;

exports[`keepLineBreaksPlugin absent does not keep line breaks between the items in an ordered list 1`] = `
<ol>
  

  <li>
    

    <p>
      item 1
    </p>
    

  </li>
  

  <li>
    

    <p>
      item 2
    </p>
    

  </li>
  

  <li>
    

    <p>
      item 3
    </p>
    

  </li>
  

</ol>
`;

exports[`keepLineBreaksPlugin absent does not keep line breaks between the items in an unordered list 1`] = `
<ul>
  

  <li>
    

    <p>
      item 1
    </p>
    

  </li>
  

  <li>
    

    <p>
      item 2
    </p>
    

  </li>
  

  <li>
    

    <p>
      item 3
    </p>
    

  </li>
  

</ul>
`;

exports[`keepLineBreaksPlugin absent does not keep line breaks under a heading 1`] = `
Array [
  "Heading",
  "
",
  <p>
    a
  </p>,
]
`;

exports[`keepLineBreaksPlugin present keeps line breaks around a blockquote 1`] = `
Array [
  <p>
    a
  </p>,
  "
",
  <br />,
  "
",
  "
",
  "
",
  <br />,
  "
",
  "
",
  "
",
  <blockquote>
    

    <p>
      b
    </p>
    

  </blockquote>,
  "
",
  <br />,
  "
",
  "
",
  "
",
  <br />,
  "
",
  "
",
  "
",
  <p>
    c
  </p>,
]
`;

exports[`keepLineBreaksPlugin present keeps line breaks around a code block 1`] = `
Array [
  <p>
    a
  </p>,
  "
",
  <br />,
  "
",
  "
",
  "
",
  <br />,
  "
",
  "
",
  "
",
  <p>
    <code>
      b
    </code>
  </p>,
  "
",
  <br />,
  "
",
  "
",
  "
",
  <br />,
  "
",
  "
",
  "
",
  <p>
    c
  </p>,
]
`;

exports[`keepLineBreaksPlugin present keeps line breaks around a horizontal rule 1`] = `
Array [
  <p>
    a
  </p>,
  "
",
  <br />,
  "
",
  "
",
  "
",
  <br />,
  "
",
  "
",
  "
",
  "
",
  <br />,
  "
",
  "
",
  "
",
  <br />,
  "
",
  "
",
  "
",
  <p>
    b
  </p>,
]
`;

exports[`keepLineBreaksPlugin present keeps line breaks around a strikethrough 1`] = `
Array [
  <p>
    a
  </p>,
  "
",
  <br />,
  "
",
  "
",
  "
",
  <br />,
  "
",
  "
",
  "
",
  <p>
    ~~xxx~~
  </p>,
  "
",
  <br />,
  "
",
  "
",
  "
",
  <br />,
  "
",
  "
",
  "
",
  <p>
    b
  </p>,
]
`;

exports[`keepLineBreaksPlugin present keeps line breaks around a table 1`] = `
Array [
  <p>
    a
  </p>,
  "
",
  <br />,
  "
",
  "
",
  "
",
  <br />,
  "
",
  "
",
  "
",
  <p>
    | a | b  |  c |  d  |
| - | :- | -: | :-: |
| a | b  |  c |  d  |
  </p>,
  "
",
  <br />,
  "
",
  "
",
  "
",
  <br />,
  "
",
  "
",
  "
",
  <p>
    c
  </p>,
]
`;

exports[`keepLineBreaksPlugin present keeps line breaks between paragraphs 1`] = `
Array [
  <p>
    a
  </p>,
  "
",
  <br />,
  "
",
  "
",
  "
",
  <br />,
  "
",
  "
",
  "
",
  <p>
    b
  </p>,
  "
",
  <br />,
  "
",
  "
",
  "
",
  <br />,
  "
",
  "
",
  "
",
  <p>
    c
  </p>,
]
`;

exports[`keepLineBreaksPlugin present keeps line breaks between the items in an ordered list 1`] = `
<ol>
  

  <li>
    

    <p>
      item 1
    </p>
    

  </li>
  

  <br />
  

  

  

  <br />
  

  

  

  <li>
    

    <p>
      item 2
    </p>
    

  </li>
  

  <br />
  

  

  

  <br />
  

  

  

  <li>
    

    <p>
      item 3
    </p>
    

  </li>
  

</ol>
`;

exports[`keepLineBreaksPlugin present keeps line breaks between the items in an unordered list 1`] = `
<ul>
  

  <li>
    

    <p>
      item 1
    </p>
    

  </li>
  

  <br />
  

  

  

  <br />
  

  

  

  <li>
    

    <p>
      item 2
    </p>
    

  </li>
  

  <br />
  

  

  

  <br />
  

  

  

  <li>
    

    <p>
      item 3
    </p>
    

  </li>
  

</ul>
`;

exports[`keepLineBreaksPlugin present keeps line breaks under a heading 1`] = `
Array [
  "Heading",
  "
",
  <br />,
  "
",
  "
",
  "
",
  <br />,
  "
",
  "
",
  "
",
  <p>
    a
  </p>,
]
`;

exports[`renderText handles the special case where there are pronouns in the name 1`] = `
<p>
  hey, 
  <span
    className="str-chat__message-mention"
    data-user-id="john"
  >
    @John (they/them)
  </span>
  , how are you?
</p>
`;

exports[`renderText handles the special case where there is a backslash in the name 1`] = `
<p>
  hey, 
  <span
    className="str-chat__message-mention"
    data-user-id="john"
  >
    @John\\Cena
  </span>
  , how are you?
</p>
`;

exports[`renderText handles the special case where there is a forward slash in the name 1`] = `
<p>
  hey, 
  <span
    className="str-chat__message-mention"
    data-user-id="john"
  >
    @John/Cena
  </span>
  , how are you?
</p>
`;

exports[`renderText handles the special case where there's at least one mention and @ symbol at the end 1`] = `
<p>
  <span
    className="str-chat__message-mention"
    data-user-id="id-username@email.com"
  >
    @username@email.com
  </span>
   @
</p>
`;

exports[`renderText handles the special case where user name matches to an e-mail pattern - 1 1`] = `
<p>
  Hello 
  <span
    className="str-chat__message-mention"
    data-user-id="id-username@email.com"
  >
    @username@email.com
  </span>
  , is 
  <a
    className=""
    href="mailto:username@email.com"
    rel="nofollow noreferrer noopener"
    target="_blank"
  >
    username@email.com
  </a>
   your @primary e-mail?
</p>
`;

exports[`renderText handles the special case where user name matches to an e-mail pattern - 2 1`] = `
<p>
  <a
    className=""
    href="mailto:username@email.com"
    rel="nofollow noreferrer noopener"
    target="_blank"
  >
    username@email.com
  </a>
   
  <span
    className="str-chat__message-mention"
    data-user-id="id-username@email.com"
  >
    @username@email.com
  </span>
   is this the right address?
</p>
`;

exports[`renderText handles the special case where user name matches to an e-mail pattern - 3 1`] = `
<p>
  <span
    className="str-chat__message-mention"
    data-user-id="id-username@email.com"
  >
    @username@email.com
  </span>
   
  <span
    className="str-chat__message-mention"
    data-user-id="id-username@email.com"
  >
    @username@email.com
  </span>
   
  <span
    className="str-chat__message-mention"
    data-user-id="id-username@email.com"
  >
    @username@email.com
  </span>
   
  <span
    className="str-chat__message-mention"
    data-user-id="id-username@email.com"
  >
    @username@email.com
  </span>
</p>
`;

exports[`renderText handles the special case where user name matches to an e-mail pattern - 4 1`] = `
<p>
  <span
    className="str-chat__message-mention"
    data-user-id="id-username@email.com"
  >
    @username@email.com
  </span>
   
  <span
    className="str-chat__message-mention"
    data-user-id="id-username@email.com"
  >
    @username@email.com
  </span>
   
  <a
    className=""
    href="mailto:username@email.com"
    rel="nofollow noreferrer noopener"
    target="_blank"
  >
    username@email.com
  </a>
   
  <span
    className="str-chat__message-mention"
    data-user-id="id-username@email.com"
  >
    @username@email.com
  </span>
</p>
`;

exports[`renderText renders custom mention 1`] = `
<p>
  <span
    className="my-mention"
    data-node-mentionedUser-id="id-username@email.com"
  >
    @username@email.com
  </span>
   
  <span
    className="my-mention"
    data-node-mentionedUser-id="id-username@email.com"
  >
    @username@email.com
  </span>
   
  <a
    className=""
    href="mailto:username@email.com"
    rel="nofollow noreferrer noopener"
    target="_blank"
  >
    username@email.com
  </a>
   
  <span
    className="my-mention"
    data-node-mentionedUser-id="id-username@email.com"
  >
    @username@email.com
  </span>
</p>
`;

exports[`renderText renders remark-gfm list and strikethrough correctly 1`] = `
Array [
  <p>
    Pick a time to meet:
  </p>,
  "
",
  <ul>
    

    <li>
      Wednesday
    </li>
    

    <li>
      Thursday
    </li>
    

    <li>
      <del>
        Sunday
      </del>
    </li>
    

    <li>
      ~Monday~
    </li>
    

  </ul>,
]
`;

exports[`renderText renders standard markdown text 1`] = `
<p>
  Hi, shall we meet on 
  <strong>
    Tuesday
  </strong>
  ?
</p>
`;
